src/rulesets/Base/array.jl

```diff
@@ -527,3 +527,42 @@ function _extrema_dims(x, dims)
     end
     return y, extrema_pullback_dims
 end
+
+#####
+##### `pmap`
+#####
+
+using UUIDs # of course we don't want this;
+            # what is a better way of uniquely identifying function calls from the forward pass?
+
+save_backs = Dict() # could this be a stack instead?
+
+# Now that there is a backwards rule for zip (albeit only in Zygote),
+# it should be fine to deal with only a single collection c
+function rrule(config::RuleConfig{>:HasReverseMode}, ::typeof(pmap), f, p::AbstractWorkerPool, c; kwargs...)
+    key = uuid1()
+    function forw(i, a)
+        y, back = rrule_via_ad(config, f, a)
+        save_backs[(key,i)] = back;
+        return y, myid()
+    end
+
+    I = reshape(eachindex(c), size(c))
+    ys_IDs = pmap(forw, I, c; kwargs...)
```

Comment:
I think the claim is that doing `pmap(a -> rrule_via_ad(config, f, a), c)` makes an array on the head process. Is there a simple way to check this, and to benchmark things?

I also wonder: for what kind of `f` does this matter? Should this maximal solution come after you've checked that there are't easier ones, somehow? e.g. for sufficiently pure `f`.

Reply:
`f` may contain data so it's possibly better to avoid serializing it often. We found this issue in Dagger.jl as well, so it seems better not to assume that serializing `f` is free.

Reply:
The issue this rule resolves is when the pullback function (`back`) is a closure over a lot of data, significantly more than the size of `y`. Roughly speaking, I think this should happen when `f` contains several non-linear steps so that a lot of data needs to be remembered for the backward pass. This can happen with completely pure `f`, so it may be difficult to check?

Reply:
Yes, I guess pure isn't it, as you say.

One check is whether `ChainRulesCore.derivatives_given_output` returns something, in which case you need not store the pullback at all. But perhaps something as cheap as `pmap(log, x)` is something nobody will ever do.

How can you test whether this copy back has happened?